File: java-api/src/main/java/org/hchery/treeleaf/db/log/LoginLog.java
package org.hchery.treeleaf.db.log;

import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.ToString;
import org.hchery.treeleaf.db.ImmutableDbModel;
import org.hchery.treeleaf.db.authentication.AuthenticationType;
import org.hchery.treeleaf.db.authentication.LoginErrorType;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.Field;

import java.io.Serial;
import java.util.Date;

/**
 * DATE: 2024/6/1
 * AUTHOR: hchery
 * URL: https://github.com/hchery
 * EMAIL: h.chery@qq.com
 */
@EqualsAndHashCode(callSuper = true)
@Data
@ToString(callSuper = true)
@Document("logs_login")
public class LoginLog extends ImmutableDbModel {
    @Serial
    private static final long serialVersionUID = -4298532843113416357L;

    @Field("user_id")
    private String userId;

    @Field("ip_addr")
    private String ipAddr;

    @Field("ip_region")
    private String ipRegion;

    @Field("user_agent")
    private String userAgent;

    @Field("login_time")
    private Date loginTime;

    @Field("authentication_type")
    private AuthenticationType authenticationType;

    @Field("success")
    private boolean success;

    @Field("error_type")
    private LoginErrorType errorType;
}

File: tree-leaf/src/main/kotlin/org/hchery/treeleaf/service/authentication/login/listener/LoginLog.kt
package org.hchery.treeleaf.service.authentication.login.listener

import org.hchery.treeleaf.components.locate.IpRegion
import org.hchery.treeleaf.db.authentication.LoginErrorType
import org.hchery.treeleaf.db.log.LoginLog
import org.hchery.treeleaf.service.authentication.login.LoginEvent
import org.hchery.treeleaf.service.log.login.LoginLogService
import org.springframework.context.ApplicationListener
import org.springframework.scheduling.annotation.Async
import org.springframework.stereotype.Component

/**
 * DATE: 2024/6/1
 * AUTHOR: hchery
 * URL: https://github.com/hchery
 * EMAIL: h.chery@qq.com
 */

@Component
class LoginEventLogListener(
    private val loginLogService: LoginLogService,
    private val ipRegion: IpRegion
) : ApplicationListener<LoginEvent> {

    @Async
    override fun onApplicationEvent(event: LoginEvent) {
        val log = toLog(event)
        loginLogService.save(log)
    }

    private fun toLog(event: LoginEvent): LoginLog {
        val log = LoginLog()
        log.userId = event.userId
        log.ipAddr = event.remoteIp
        log.ipRegion = ipRegion(event.remoteIp)
        log.userAgent = event.userAgent
        log.loginTime = event.time
        log.isSuccess = event.loginSuccess
        log.authenticationType = event.authenticationType
        log.errorType = getErrorType(event)
        return log
    }

    private fun getErrorType(event: LoginEvent): LoginErrorType? {
        if (event.loginSuccess) {
            return null
        }
        return event.loginErrorType ?: LoginErrorType.SERVER_ERROR
    }
}
